ql/src/java/org/apache/hadoop/hive/ql/optimizer/ExtendParentReduceSinkOfMapJoin.java

```diff
@@ -0,0 +1,205 @@
+      // Therefore, RS's writer and MapJoin's reader should have the same TableDesc. We create valueTableDesc
+      // here again because it can be different from RS's valueSerializeInfo due to ColumnPruner.
+      List<TableDesc> newMapJoinValueFilteredTableDescs = new ArrayList<>();
+      for (byte pos = 0; pos < mapJoinOp.getParentOperators().size(); pos++) {
+        TableDesc tableDesc;
+
+        if (pos == mapJoinDesc.getPosBigTable() || filterMap[pos] == null) {
+          // We did not change corresponding parent operator. Use the original tableDesc.
+          tableDesc = mapJoinDesc.getValueFilteredTblDescs().get(pos);
+        } else {
+          // Create a new TableDesc based on corresponding parent RSOperator.
+          ReduceSinkOperator parent = (ReduceSinkOperator) mapJoinOp.getParentOperators().get(pos);
+          ReduceSinkDesc pRsConf = parent.getConf();
+
+          tableDesc =
+              PlanUtils.getMapJoinValueTableDesc(
+                  PlanUtils.getFieldSchemasFromColumnList(pRsConf.getValueCols(), "mapjoinvalue"));
+        }
+
+        newMapJoinValueFilteredTableDescs.add(tableDesc);
+      }
+      mapJoinDesc.setValueFilteredTblDescs(newMapJoinValueFilteredTableDescs);
+    }
+
+    return null;
+  }
+
+  private ExprNodeDesc generateFilterTagExpression(int[] filterMap, List<ExprNodeDesc> filterExprs) {
+    ExprNodeDesc filterTagExpr = new ExprNodeConstantDesc(shortType, (short) 0);
+    Map<Byte, ExprNodeDesc> filterExprMap = getFilterExprMap(filterMap, filterExprs);
+
+    for (Map.Entry<Byte, ExprNodeDesc> entry: filterExprMap.entrySet()) {
+      ExprNodeDesc filterTagMaskExpr = generateFilterTagMask(entry.getKey(), entry.getValue());
+
+      if (filterTagExpr instanceof ExprNodeConstantDesc) {
+        filterTagExpr = filterTagMaskExpr;
+      } else {
+        List<ExprNodeDesc> plusArgs = new ArrayList<>(2);
+        plusArgs.add(filterTagMaskExpr);
+        plusArgs.add(filterTagExpr);
+        filterTagExpr = new ExprNodeGenericFuncDesc(shortType, new GenericUDFOPPlus(), plusArgs);
+      }
+    }
+
+    return filterTagExpr;
+  }
+
+  private Map<Byte, ExprNodeDesc> getFilterExprMap(int[] filterInfo, List<ExprNodeDesc> filterExprs) {
+    Map<Byte, ExprNodeDesc> filterExprMap = new HashMap<>();
+
+    int exprListOffset = 0;
+    for (int idx = 0; idx < filterInfo.length; idx = idx + 2) {
+      byte tag = (byte) filterInfo[idx];
+      int length = filterInfo[idx + 1];
+
+      int nextExprOffset = exprListOffset + length;
+      List<ExprNodeDesc> andArgs = filterExprs.subList(exprListOffset, nextExprOffset);
+      exprListOffset = nextExprOffset;
+
+      if (andArgs.size() == 1) {
+        filterExprMap.put(tag, andArgs.get(0));
+      } else if (andArgs.size() > 1) {
+        filterExprMap.put(tag, ExprNodeDescUtils.and(andArgs));
+      }
+    }
+
+    return filterExprMap;
+  }
+
+  private ExprNodeDesc generateFilterTagMask(byte tag, ExprNodeDesc condition) {
```

Comment:
Could you please elaborate the purpose of the tag mask and how it is calculated?

Reply:
I added a comment that explains the computation performed by the resultant ExprNodeDesc.

Reply:
I tried to understand how`JoinUtil.java.isFiltered` works.
IIUC `0` means the row can be forwarded. A value greater than `0` means it should be filtered out.
The mask is used to represent by which expressions were false.
Is that correct?

Reply:
The meaning of value is correct. filterTag == 0 means that the row satisfies all join conditions and will be forwarded. If filterTag is greater than 0, the row should be filtered out unless there is an outer join.

The meaning of mask is a bit complicated; it indicates the alias of table that associates with the failed filter expression. (alias is also called as tag or pos.)

For example, suppose that we are joining 3 tables A, B, and C using the following SQL command.
(Assume that alias 0 = table A, alias 1 = table B, and alias 2 = table C)
```
SELECT * FROM A JOIN B ON A.k = B.k and A.k > 10 and A.v = 0 JOIN C ON A.k = C.k AND A.k < 20;
```
There are 3 filter expressions in the above command: `A.k > 10`, `A.v = 0`, and `A.k < 20`. You can see that all of them are evaluated on table A, but they do not belong to the same join relations: `A.k > 10`, `A.v = 0` belong to `A JOIN B` while `A.k < 20` belongs to `A JOIN C`.
The mask is used to distinguish `A JOIN B` and `A JOIN C` in this situation. If filterTag & 0x04 == 1, i.e. the third lowest bit of filterTag is 1, then one of the filter expressions that associates with alias 2 returns false. In our case, alias 2 means table C. So the filterTag & 0x04 == 1 means that `A.k < 20` returns false.

filterTag is used for properly computing outer joins. You can refer to CommonJoinOperator.genObject(), which uses getFilterTag() and plays a central role in the join algorithm.